src/coreclr/nativeaot/Runtime/gcrhenv.cpp

```diff
@@ -212,74 +212,104 @@ bool RedhawkGCInterface::InitializeSubsystems()
 }
 #endif // !DACCESS_COMPILE
 
-void* GcAllocInternal(EEType *pEEType, uint32_t uFlags, uintptr_t cbSize, Thread* pThread)
+Object* GcAllocInternal(EEType *pEEType, uint32_t uFlags, uintptr_t numElements, Thread* pThread)
 {
     ASSERT(!pThread->IsDoNotTriggerGcSet());
 
-    if (cbSize >= RH_LARGE_OBJECT_SIZE)
+    size_t cbSize = pEEType->get_BaseSize();
+
+    if (pEEType->get_ComponentSize() != 0)
```

Comment:
There are a few extra if-checks and an extra multiplication compared to what it used to be, but it is so much simpler and with less assembly code to have everything centralized in one method.